### PR TITLE
Feature: move delete records to a role_action instead of an enterprise_setting

### LINKED ISSUE #5379: Make deleting records (invoices/cash payments/vouchers) into a Role Action
We have certain "role actions" that allow the users in possession of this role to do super user things.  For example, unposting transactions.  Here is the presentation of the modal:
image

I propose we make the current enterprise setting for `enable_delete_records` into a Role Action.  It is currently an enterprise setting:
https://github.com/IMA-WorldHealth/bhima/blob/1f0e637814c67cc04f9ac0ae99744e5a16d7f8b4/server/controllers/admin/enterprises.js#L47

This would do two things:

1. It would allow you to have some users (like an IMA user) have the permission to delete records _without_ turning it on for everyone else.  This may be very useful during training exercises, for example.
2. It would elevate the privileges required to toggle this setting from "anyone who has access to the enterprise module" to "anyone who has access to roles".  This is presumably a stricter setting.

### PR DESCRIPTION
**Pull request by jniles**
This commit moves the delete records value to a role_action instead of an enterprise_setting.  This change makes it so that the deletion ability is partitioned by user instead of just being globally "on" or "off" for an entire enterprise.

Further, we've also separated out each type of deletion so that an enterprise can choose to allow the deletion of only a certain type of transaction (for example, purchase orders) without allowing any others.  Alternatively, they can give some users some permissions but not all users all permissions.

Here is what it looks like:

rod4ARuPsJ

How to tests this:
  1. Build the test database.
  2. Sign in as the superuser
  3. Go to the invoice registry
  4. Observe that you have access to delete records in the dropdown menu.
  5. Navigate to the posting journal.
  6. Observe that you have access to the "unpost" button at the top of the journal.
  7. Navigate to the roles module. 
  8. Edit the superuser **actions** and uncheck the "Can unpost transactions" option and the "Delete Invoice" options.
  9. Submit.
  9. Navigate to both modules and observe that you no longer have access.
  10. Navigate back and check the options again
  11. You will now have access.

Closes  #5379.

### CODE REVIEW

File: client/src/js/constants/bhConstants.js
Comment (jmcameron) [inline client/src/js/constants/bhConstants.js:20]: I wonder whether we should consider using non-sequential numbers such as:
      CAN_EDIT_ROLES : 1
      CAN_UNPOST_TRANSACTIONS : 101,
      DELETE_CASH_PAYMENT : 201,
      DELETE_INVOICE : 202,
      DELETE_PURCHASE_ORDER : 203,
      DELETE_STOCK_MOVEMENT : 204,
      DELETE_VOUCHER : 205,

So that new items can be added in the future without requiring numbering or having strange numbering sequences.   Since we are just using these as constants, the fact that they are non-sequential should not matter.

On second thought... I noticed later that these are actually the record numbers in an SQL table, so it is probably better to keep them sequential.

File: client/src/modules/invoices/registry/templates/action.cell.html
Comment (jmcameron) [inline client/src/modules/invoices/registry/templates/action.cell.html:68]: Just a stylistic question:
- Might it be better to keep the permissions function allowsRecordDeletion() and hide the bh-has-permission, bhConstants.actions, etc hidden from the code.  That way if we decide to change that in the future, we only have to change the function.  I presume the function would be available in the Session object so it would be shared by all potential users?

**Review by jmcameron — APPROVED** [review]
The code changes look good.

I ran tests, as suggested, and it works as advertised.

I had one question (in general regarding permissions, and not blocking this PR): It looks like the permissions mechanism prevents the non-permitted link/button from being displayed.  Is it still possible for someone to fake the request (eg, with injected client Javascript) and perform the action anyway?  In other words, is there a mechanism on the server side preventing non-permitted actions even if they slip by the client side?  (I skimmed the server side code changes and did see anything...)

File: client/src/js/constants/bhConstants.js
Comment (jniles) [reply client/src/js/constants/bhConstants.js:20]: This is a good thought, but at this point there aren't enough people smashing in permissions to make it worth coming up with a system.  This would make a lot more sense for our `unit` table, actually.  The probably is that we'd need to deal with legacy installations...

File: client/src/modules/invoices/registry/templates/action.cell.html
Comment (jniles) [reply client/src/modules/invoices/registry/templates/action.cell.html:68]: This isn't just stylistic - it is encapsulation so we don't repeat ourselves.  What the `bhHasPermission` does is encapsulates the `Session` check so the parent controller (Invoice Registry, Cash Registry, etc) doesn't have to contain any permissions logic.  It only imports Constants to pass to the directive.  To use `allowsRecordDeletion()`, we would need to import the services that are requred to figure out if record deletion is allowed.

One way that we could make this "simpler" would be to assign each action a `data-*` attribute that we affix to the DOM node.  Or we could just pass a string to the directive.  We then we wouldn't even need import the `bhConstants` into the parent controller, and the parent controller can be only concerned with the Invoice Registry, not users and permissions.  However, I judged (and may be wrong) that this is more things to break if we decide to change logic.

Here is how that would work:
```html
<!-- using a data-* attribute -->
<a bh-has-permission data-delete-invoice>
  <!-- all the code -->
</a>

<!-- using a string on the directive-->
<a bh-has-permission="delete-invoice">
  <!-- all the code -->
</a>
```

To address your second point, the functionality shipped to the client (Session, etc) is only for that single user/machine.  So, it shouldn't be "shared" in the sense of sharing state across users and devices, but it is indeed shipped to all clients.

**Comment by jniles** [comment]
@jmcameron yes, there are no server-side checks.  That is a good idea.  It shouldn't be hard to add them.  Let me do so before merging this.

**Comment by jniles** [comment]
@jmcameron this is ready for a second look.

File: server/controllers/finance/purchases.js
Comment (jmcameron) [inline server/controllers/finance/purchases.js]: Don't you want to get this from bhConstants? (DRY: Don't Repeat Yourself)

File: server/controllers/finance/purchases.js
Comment (jmcameron) [inline server/controllers/finance/purchases.js:615]: Nice!

File: server/controllers/finance/transactions.js
Comment (jmcameron) [inline server/controllers/finance/transactions.js]: Again, shouldn't these be from bhConstants?

File: server/controllers/finance/transactions.js
Comment (jmcameron) [inline server/controllers/finance/transactions.js:109]: Nice!

File: server/controllers/stock/index.js
Comment (jmcameron) [inline server/controllers/stock/index.js]: Here too...

File: server/models/bhima.sql
Comment (jmcameron) [inline server/models/bhima.sql:342]: May want to add a note that these 'id's must match the ones in bhConstants.

File: server/models/migrations/next/migrate.sql
Comment (jmcameron) [inline server/models/migrations/next/migrate.sql:83]: And here too.

**Review by jmcameron — COMMENTED** [review]
I tested deleting an invoice.  The permissions worked perfectly.  However (when I had permission) and tried to delete an invoice it complained that a transaction was referencing it so it could not be deleted.  That was good and to be expected.  What I wanted to note is that the red warning banner on the top of the screen from this error stays there when I switch to other pages.  When I did a hard refresh it disappeared.   Also, the error message is rather vague about what transaction to delete.  Since I assume only experienced operators will be doing this, this may not be a problem since they would know how to figure that out.  But should we consider adding more details to the error message?

The posting journal "unpost" works as advertised, although I did not actually try it.

File: server/controllers/finance/transactions.js
Comment (jniles) [reply server/controllers/finance/transactions.js]: bhConstants is a client-side script, so no.  We probably should assemble these from a constant script on the server.

https://github.com/IMA-WorldHealth/bhima/issues/1221 requests that we build out all our constants from a central location.  This will require a pretty significant refactor.  I'll make this change the start, but we can later go through and update account type ids, etc, that are hard-coded.

File: server/controllers/finance/transactions.js
Comment (jmcameron) [inline server/controllers/finance/transactions.js:21]: I like it!

File: server/config/constants.js
Comment (jmcameron) [inline server/config/constants.js:1]: Excellent!

**Review by jmcameron — APPROVED** [review]
Looks great!
Tested and still works...

**Pull request merged**
